mineral-ui/mineral-ui
# Legacy React lifecycle migration - Mineral components, excluding Table

### Description

Refactors most Mineral components to no longer use legacy React lifecycle methods.

* Refactor our `createRootNode` (and similar - e.g. `getItems`, `getControlNode`) patterns to use a single level memoization cache instead of Reacts deprecated `componentWillUpdate` lifecycle method.
* Update Link component to use `createRootNode` pattern
* Extract `mountInWrapper` test util and update a couple of tests to use it.  This was more important in my original branch which included memoization logic tests. 
* Add safe check to `Menu.isGroupedData`
* Moves root ThemeProvider from index.js to App.js - attempting to keep index as vanilla as possible.

### Motivation and context

Closes #763 #722 

Note that this work does not complete the migration as is part of an epic #652

### Screenshots, videos, or demo, if appropriate

http://763-lifecycle--mineral-ui.netlify.com

### How to test

Existing automated tests plus manual testing as you see fit.

I tried several means of writing additional tests for the memoization logic, but the only way that I could successfully do that was to place the getRootNode, etc. methods as static functions on the components, and that didn't feel right as it is essentially part of the public API.  I ideally wanted them to be separate named exports that could then be mocked in unit tests.  I know that we really don't need to bother testing private methods, but when implementing these changes as well as experimenting with refactoring Table's more complex logic, they proved quite valuable.  That work can be found on the #652-lifecycle branch and I also have additional mocking work locally. A good discussion on jest mocking strategies can be found at https://github.com/facebook/jest/issues/936.

### Types of changes

- Other (provide details below)

Refactor

### Checklist

* [x] Automated tests written and passing - **mostly existing tests**
